src/read.rs

```diff
@@ -223,26 +324,65 @@ impl<R: Read + io::Seek> ZipArchive<R> {
         self.names_map.keys().map(|s| s.as_str())
     }
 
+    /// Search for a file entry by name, decrypt with given password
+    pub fn by_name_decrypt<'a>(
+        &'a mut self,
+        name: &str,
+        password: &[u8],
+    ) -> ZipResult<ZipFile<'a>> {
+        self.by_name_internal(name, Some(password))
+    }
+
     /// Search for a file entry by name
     pub fn by_name<'a>(&'a mut self, name: &str) -> ZipResult<ZipFile<'a>> {
+        self.by_name_internal(name, None)
+    }
+
+    fn by_name_internal<'a>(
+        &'a mut self,
+        name: &str,
+        password: Option<&[u8]>,
+    ) -> ZipResult<ZipFile<'a>> {
         let index = match self.names_map.get(name) {
             Some(index) => *index,
             None => {
                 return Err(ZipError::FileNotFound);
             }
         };
-        self.by_index(index)
+        self.by_index_internal(index, password)
+    }
+
+    /// Get a contained file by index, decrypt with given password
+    pub fn by_index_decrypt<'a>(
+        &'a mut self,
+        file_number: usize,
+        password: &[u8],
+    ) -> ZipResult<ZipFile<'a>> {
+        self.by_index_internal(file_number, Some(password))
     }
 
     /// Get a contained file by index
     pub fn by_index<'a>(&'a mut self, file_number: usize) -> ZipResult<ZipFile<'a>> {
+        self.by_index_internal(file_number, None)
+    }
+
+    fn by_index_internal<'a>(
+        &'a mut self,
+        file_number: usize,
+        mut password: Option<&[u8]>,
+    ) -> ZipResult<ZipFile<'a>> {
         if file_number >= self.files.len() {
             return Err(ZipError::FileNotFound);
         }
-        let data = &mut self.files[file_number];
+        let ref mut data = self.files[file_number];
 
-        if data.encrypted {
-            return unsupported_zip_error("Encrypted files are not supported");
+        if password == None {
+            if data.encrypted {
```

Comment:
I think this would be easier to follow like this:
```rust
match (password, data.encrypted) {
  (None, true) => return Err(..),
  (Some(_), false) => password = None,
  _ => {}
}
```

Reply:
Back when I wrote this code, I didn't even know this was possible with matcher clauses (I only learned that a week ago). Looks elegant, I applied that change.